File: src/main/java/random/telegramhomebot/controllers/HostCsvController.java
package random.telegramhomebot.controllers;

import org.apache.commons.collections4.CollectionUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.multipart.MultipartFile;
import random.telegramhomebot.model.Host;
import random.telegramhomebot.repository.HostRepository;
import random.telegramhomebot.services.HostsCsvService;

import javax.annotation.Resource;
import javax.servlet.http.HttpServletResponse;
import java.lang.invoke.MethodHandles;
import java.util.List;

import static random.telegramhomebot.AppConstants.Hosts.HOSTS_MAPPING;
import static random.telegramhomebot.AppConstants.Hosts.REDIRECT_HOSTS;
import static random.telegramhomebot.AppConstants.HostsCsv.FILE_REQ_PARAM;
import static random.telegramhomebot.AppConstants.HostsCsv.HOSTS_CSV_EXPORT_MAPPING;
import static random.telegramhomebot.AppConstants.HostsCsv.HOSTS_CSV_IMPORT_MAPPING;

@Controller
@RequestMapping(HOSTS_MAPPING)
public class HostCsvController {

	private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());

	@Resource
	private HostRepository hostRepository;
	@Resource
	private HostsCsvService hostsCsvService;

	@GetMapping(HOSTS_CSV_EXPORT_MAPPING)
	public void exportHostsToCsvFile(HttpServletResponse response) throws Exception {
		hostsCsvService.exportHostsToCsvFile(response, hostRepository.findAll());
	}

	@PostMapping(HOSTS_CSV_IMPORT_MAPPING)
	public String importHostsFromCsvFile(@RequestParam(FILE_REQ_PARAM) MultipartFile file, Model model) {

		if (file.isEmpty()) {
			// TODO: impl error message
			log.error("File to import is empty!");
//			model.addAttribute("message", "Please select a CSV file to upload.");
		} else {
			try {
				List<Host> hostsToImport = hostsCsvService.parseHostsFromCsvFile(file);
				if (CollectionUtils.isNotEmpty(hostsToImport)) {
					hostRepository.saveAll(hostsToImport);
				}
			} catch (Exception e) {
				// TODO: impl error message
//				model.addAttribute("message", "An error occurred while processing the CSV file.");
				log.error(e.getMessage(), e);
			}
		}

		return REDIRECT_HOSTS;
	}
}

File: src/main/java/random/telegramhomebot/controllers/HostCsvController.kt
package random.telegramhomebot.controllers

import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.ResponseBody
import org.springframework.web.multipart.MultipartFile
import random.telegramhomebot.const.AppConstants.FILE_REQ_PARAM
import random.telegramhomebot.const.AppConstants.HOSTS_CSV_EXPORT_MAPPING
import random.telegramhomebot.const.AppConstants.HOSTS_CSV_IMPORT_MAPPING
import random.telegramhomebot.const.AppConstants.HOSTS_MAPPING
import random.telegramhomebot.services.csv.HostsCsvService
import random.telegramhomebot.services.hosts.HostService
import random.telegramhomebot.utils.logger
import javax.servlet.http.HttpServletResponse

@Controller
@RequestMapping(HOSTS_MAPPING)
class HostCsvController(
    private val hostService: HostService,
    private val hostsCsvService: HostsCsvService
) {
    val log = logger()

    @GetMapping(HOSTS_CSV_EXPORT_MAPPING)
    @Throws(Exception::class)
    fun exportHostsToCsvFile(response: HttpServletResponse) {
        hostsCsvService.exportHostsToCsvFile(response)
    }

    @ResponseBody
    @PostMapping(HOSTS_CSV_IMPORT_MAPPING)
    fun importHostsFromCsvFile(
        @RequestParam(FILE_REQ_PARAM) file: MultipartFile
    ): ResponseEntity<String> {
        if (file.isEmpty) {
            log.error("File to import is empty!")
            return ResponseEntity.badRequest().body("Please select a CSV file to upload.")
        }
        return try {
            val hostsToImport = hostsCsvService.parseHostsFromCsvFile(file)
            if (hostsToImport.isNotEmpty()) {
                hostService.saveAllHosts(hostsToImport)
            }
            ResponseEntity.ok().build()
        } catch (e: Exception) {
            val err = "An error occurred while processing the CSV file."
            log.error(e.message, e)
            ResponseEntity.badRequest().body(err)
        }
    }
}
